src/Middleware/RequestDecompression/src/SizeLimitedStream.cs

```diff
@@ -0,0 +1,101 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.RequestDecompression;
+
+internal class SizeLimitedStream : Stream
+{
+    private readonly Stream _innerStream;
+    private readonly long? _sizeLimit;
+
+    private long _totalBytesRead;
+
+    public SizeLimitedStream(Stream innerStream, long? sizeLimit)
+    {
+        if (innerStream is null)
+        {
+            throw new ArgumentNullException(nameof(innerStream));
+        }
+
+        _innerStream = innerStream;
+        _sizeLimit = sizeLimit;
+    }
+
+    public override bool CanRead => _innerStream.CanRead;
+
+    public override bool CanSeek => false;
+
+    public override bool CanWrite => false;
+
+    public override long Length
+    {
+        get { throw new NotSupportedException(); }
+    }
+
+    public override long Position
+    {
+        get { throw new NotSupportedException(); }
+        set { throw new NotSupportedException(); }
+    }
+
+    public override void Flush()
+    {
+        throw new NotSupportedException();
+    }
+
+    public override int Read(byte[] buffer, int offset, int count)
```

Comment:
EnableBuffering could be used before or after, it depends on what you're trying to do. It's true that the decompression stream wouldn't be seekable after this middleware without additional buffering.

Reply:
> It's true that the decompression stream wouldn't be seekable after this middleware without additional buffering.

I'm still not sure of my suggestion but should at least `CanSeek`, `Length` and `Position` return the `_innerStream` results instead of `false` and throwing exceptions?

Reply:
Sure, this wrapper should delegate, though the decompression streams will do the same thing.